snuba/migrations/policies.py

```diff
@@ -0,0 +1,71 @@
+from abc import ABC, abstractmethod
+
+from snuba.migrations.groups import get_group_loader
+from snuba.migrations.runner import MigrationKey, Runner
+from snuba.migrations.status import Status
+
+
+class MigrationPolicy(ABC):
+    """
+    A MigrationPolicy implements `can_run` and
+    `can_reverse` methods that determines whether or not
+    a migration can be run or reversed from snuba admin.
+
+    A policy can be used by a migration group to determine the
+    level of access a group has to run/reverse migrations.
+    """
+
+    @abstractmethod
+    def can_run(self, migration_key: MigrationKey) -> bool:
+        raise NotImplementedError
+
+    @abstractmethod
+    def can_reverse(self, migration_key: MigrationKey) -> bool:
+        raise NotImplementedError
+
+
+class ReadOnlyPolicy(MigrationPolicy):
+    """
+    No migration is allowed to be run or reversed.
+    """
+
+    def can_run(self, migration_key: MigrationKey) -> bool:
+        return False
+
+    def can_reverse(self, migration_key: MigrationKey) -> bool:
+        return False
+
+
+class WriteSafeAndPendingPolicy(MigrationPolicy):
+    """
+    Migrations can be run that are non-blocking, as determined
+    by the `blocking` attribute set on a migration.
+
+    Reversing a migration is considered blocking by this policy,
+    and only migrations in a pending state can be reversed.
+    """
+
+    def can_run(self, migration_key: MigrationKey) -> bool:
+        migration = get_group_loader(migration_key.group).load_migration(
```

Comment:
What's the interaction of migration policies and groups? Since we are locking down access by group doesn't the policy need to check for access to the group before loading the migration?

Reply:
Do you think the policy should keep the state of the group then? And have a separate policy for every group? even if it's the same policy type

I have the interaction described at the top but you could do something like look up the mapping and then create the policy with the group `policy = WriteAll("sessions")` or whatever in the endpoint where you are actually checking the policy

Reply:
update: I added this to the docstring but the assumption is that the access to the group has been verified before checking the policy. So the policy doesn't have the responsibility of verifying the access to the group